lib/src/main/java/com/diffplug/spotless/protobuf/BufStep.java

```diff
@@ -0,0 +1,99 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.diffplug.spotless.protobuf;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.Serializable;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.List;
+import java.util.regex.Pattern;
+
+import javax.annotation.Nullable;
+
+import com.diffplug.spotless.ForeignExe;
+import com.diffplug.spotless.FormatterFunc;
+import com.diffplug.spotless.FormatterStep;
+import com.diffplug.spotless.ProcessRunner;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+public class BufStep {
+	public static String name() {
+		return "buf";
+	}
+
+	public static String defaultVersion() {
+		return "1.4.0";
+	}
+
+	private final String version;
+	private final @Nullable String pathToExe;
+
+	private BufStep(String version, @Nullable String pathToExe) {
+		this.version = version;
+		this.pathToExe = pathToExe;
+	}
+
+	public static BufStep withVersion(String version) {
+		return new BufStep(version, null);
+	}
+
+	public BufStep withPathToExe(String pathToExe) {
+		return new BufStep(version, pathToExe);
+	}
+
+	public FormatterStep create() {
+		return FormatterStep.createLazy(name(), this::createState, State::toFunc);
+	}
+
+	private State createState() throws IOException, InterruptedException {
+		String instructions = "https://docs.buf.build/installation";
+		String exeAbsPath = ForeignExe.nameAndVersion("buf", version)
+				.pathToExe(pathToExe)
+				.versionRegex(Pattern.compile("(\\S*)"))
+				.fixCantFind("Try following the instructions at " + instructions + ", or else tell Spotless where it is with {@code buf().pathToExe('path/to/executable')}")
+				.confirmVersionAndGetAbsolutePath();
+		return new State(this, exeAbsPath);
+	}
+
+	@SuppressFBWarnings("SE_TRANSIENT_FIELD_NOT_RESTORED")
+	static class State implements Serializable {
+		private static final long serialVersionUID = -1825662356883926318L;
+		// used for up-to-date checks and caching
+		final String version;
+		// used for executing
+		final transient List<String> args;
+
+		State(BufStep step, String exeAbsPath) {
+			this.version = step.version;
+			this.args = Arrays.asList(exeAbsPath, "format");
+		}
+
+		String format(ProcessRunner runner, String input, File file) throws IOException, InterruptedException {
+			String[] processArgs = args.toArray(new String[args.size() + 1]);
+			// add an argument to the end
+			processArgs[args.size()] = file.getAbsolutePath();
+			return runner.exec(input.getBytes(StandardCharsets.UTF_8), processArgs).assertExitZero(StandardCharsets.UTF_8);
```

Comment:
Bummer, we've got a problem. Your answer about the [test needing the file to exist](https://github.com/diffplug/spotless/pull/1208/files#discussion_r871996286) made me suspicious, so I did a little digging.

On this line you are passing the file's content on `stdin` as `input.getBytes(...)` but the buf command is ignoring that and just reading the file content from disk. That breaks Spotless' model - for example, any steps that get applied before `buf()` will get wiped out, and our [IDE integrations](https://github.com/diffplug/spotless/blob/main/plugin-gradle/IDE_HOOK.md) will also not work as expected.

Looks like this is a known issue, and they plan to add the ability to read `buf` from stdin: https://github.com/bufbuild/buf/issues/1035

I'd like to delay merging this until `buf` adds this feature. If this delay would be a big hassle to your workflow, I'm okay merging this semi-broken feature in, with the promise that we'll stop supporting this version of `buf` once they release a version which supports `stdin`.

Reply:
Either is fine. I can use the buf-gradle-plugin in the meantime. Bummer to not be able to use `spotlessApply` for everything!

Reply:
Since the buf-gradle-plugin is already available and the buf team has accepted the feature request, I'll wait until it gets resolved rather than merge something half-broken.

Reply:
It doesn't seem like the Buf team is going to be taking on stdin support anytime soon. What are your thoughts on merging this semi-broken feature? Would it cause problems for any common formatting use cases besides license headers?

Reply:
Oh, and since I opened this PR the Buf team has begun publishing a Buf binary to Maven Central. Is there an existing pattern for locating a binary that way?

Reply:
I'm okay merging this as-is.

> Would it cause problems for any common formatting use cases

Main thing it breaks is idempotence. Most formatters have idempotency bugs, [which Spotless quietly fixes](https://github.com/diffplug/spotless/blob/main/PADDEDCELL.md#a-misbehaving-step), and that won't work here. It also means that things like `replace`, `license`, `tabsToSpaces` etc will only work if you put them *after* the buf step, they can't come before. Not great, but we can document it and point them to the buf issue to thumbs-up it.

> publishing a Buf binary to Maven Central

I assume the binary has to be extracted from the jar? We don't have any formatters like that currently. I'd be happy to merge that in, but even happier to leave that up to some other plugin and document that in our docs. Let me know when you're ready to merge and we'll let it rip :)

Reply:
Ok - letting `buf-gradle-plugin` handle that is probably best. I corrected the docs pointing to it (Buf in-housed it earlier this year) but the instructions are the same otherwise. Extracting from the jar would mean writing code to support specifying a Maven artifact as a source for a binary and I don't think I want to tackle that at the moment given that it can be added later and we're going to put this behind common configuration code anyways.

I'm happy as this is! Should I take care of those doc notes? I checked the box to let you do it if you want to just do it quickly yourself.